include/lbann/execution_contexts/sgd_execution_context.hpp

```diff
@@ -35,8 +35,16 @@ namespace lbann {
 class sgd_termination_criteria : public termination_criteria
 {
 public:
+  sgd_termination_criteria(size_t max_num_batches,
+                           size_t max_num_epochs)
+    : termination_criteria{max_num_batches},
+      m_max_epochs{max_num_epochs}
+  {
+  }
   ~sgd_termination_criteria() = default;
-  size_t num_epochs;
+  size_t max_num_epochs() const noexcept { return m_max_epochs; }
+private:
+  size_t m_max_epochs;
```

Comment:
`sgd_execution_context` counts both. The step (batch) count is in the base `execution_context` class, since most of our training algorithms are expected to be iterative.

I strongly dislike the "epoch concept" and would love to purge it from the code.

Reply:
I think that this kind of purge is better handled when we finish reworking the data readers / data source / data coordinator.